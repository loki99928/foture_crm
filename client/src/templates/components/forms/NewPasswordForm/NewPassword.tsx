import React from "react";
import {NavLink} from "react-router-dom";
import {Form, Formik} from "formik";
import cn from "classnames"
import * as Yup from "yup";
import {useNavigate, useParams} from "react-router";
import {useDispatch} from "react-redux";

import s from "../Form.module.scss"
import bannerForm from "../../../assets/images/bg-head-form.jpg"
import {FormikControlBtn, FormikControlFields} from "../FormikControl";
import {
    IApiUsersCreateNewPasswordData,
    IApiUsersCreateNewPasswordResponse,
    ResultStatusCodeEnum
} from "../../../../types/ApiUsersTypes";
import {FormikType} from "../FormType";
import {CreatingNewPasswordApi} from "../../../../redux/Thank/Auth";

const SignupSchema = Yup.object().shape({
    password: Yup
        .string()
        .min(5, 'Too Short. Min length 5 symbol')
        .max(50, 'Too Long. Max length 50 symbol')
        // todo-dv вернуть валидацию
        // .matches( /^(?=.*[a-z])/, "Must one Lowercase" )
        // .matches( /^(?=.*[A-Z])/, "Must one Uppercase" )
        // .matches( /^(?=.*[0-9])/, "Must One Number" )
        // .matches( /^(?=.*[!@#\$%\^&\*])/, "Must One Special Case Character(@, #, $, %, ^ ,&, *)" )
        .required('Required'),
    double_password: Yup
        .string()
        .oneOf([Yup.ref('password')], 'Password mismatch')
        .required('Required')
});


export const NewPassword: React.FC = () => {

    const dispatch = useDispatch()
    const navigate = useNavigate()
    const {token} = useParams()

    const initialValues = {password: '', double_password: '', token}

    async function creatingNewPassword(values: IApiUsersCreateNewPasswordData, formikEvent: FormikType) {
        let {setSubmitting, setFieldError} = formikEvent
        setSubmitting(true);
        const res = await dispatch(CreatingNewPasswordApi(values)) as unknown as IApiUsersCreateNewPasswordResponse

        if (res.statusCode === ResultStatusCodeEnum.Success){
            navigate('/message', {
                state: {
                    type: 'newPassword'
                } })
        }
        setSubmitting(false);
    }

    const onSubmit = async (values: IApiUsersCreateNewPasswordData, {setSubmitting, setFieldError}: FormikType) => {
        setSubmitting(false);
        await creatingNewPassword(values, {setSubmitting, setFieldError})
    }

    return (
        <div className={s.blockForm}>
            <Formik
                initialValues={initialValues}
                onSubmit={onSubmit}
                validationSchema={SignupSchema}
                enableReinitialize
                validateOnBlur={true}
                validateOnChange={true}
            >
                {({
                      isSubmitting,
                      isValid,
                      errors,
                      touched,
                      values
                  }) => (
                    <Form>
                        <div className={s.formBanner}>
                            <img src={bannerForm} alt=""/>
                        </div>
                        <div className={s.containerFields}>
                            <h2 className={s.formTitle}>New password</h2>
                            <FormikControlFields
                                className={cn(s.formField)}
                                errors={errors}
                                touched={touched}
                                values={values}
                                control="input"
                                type="password"
                                label="Your password"
                                name="password"
                                htmlFor="password"
                            />
                            <FormikControlFields
                                className={cn(s.formField)}
                                errors={errors}
                                touched={touched}
                                values={values}
                                control="input"
                                type="password"
                                label="Confirm password"
                                name="double_password"
                                htmlFor="double_password"
                            />
                            <FormikControlBtn
                                control="submit"
                                label="Send"
                                type="submit"
                                disabled={isSubmitting || !isValid}
                            />
                            <div className={cn(s.formFooter, s.footer__form)}>
                                <NavLink to="/auth/">Authorize</NavLink>
                                <NavLink to="/registration/">registration</NavLink>
                            </div>
                        </div>
                    </Form>
                )}
            </Formik>
        </div>
    )
}


