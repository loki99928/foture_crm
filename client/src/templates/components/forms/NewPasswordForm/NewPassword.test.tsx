import React from "react";
import {act, fireEvent, screen} from "@testing-library/react";

import {renderWithRouter} from "../../../../helpers/test/renderWithRouter";
import {NewPassword} from "./NewPassword";

describe('NewPassword', () => {

    it('Snapshot', function () {
        const { asFragment } = renderWithRouter(<NewPassword/>)
        expect(asFragment()).toMatchSnapshot()
    })

    describe('input password', () => {

        it('have',  () => {
            renderWithRouter(<NewPassword/>)
            expect(screen.getByTestId('input_password')).toBeInTheDocument()
        })

        it('validate',  async () => {
            renderWithRouter(<NewPassword/>)
            const password = screen.getByTestId('input_password')
            await act( async () => {
                fireEvent.change(password, {
                    target: { value: '123pass456345' }
                })
            });
            await act( async () => {
                fireEvent.focusOut(password)
            });
            expect(screen.queryByTestId("formTextError")).toBeNull();
        })

        it('too short', async () => {
            renderWithRouter(<NewPassword/>)
            const password = screen.getByTestId('input_password')
            await act( async () => {
                fireEvent.change(password, {
                    target: { value: 'lk' }
                })
            });
            await act( async () => {
                fireEvent.focusOut(password)
            });
            expect(screen.queryByTestId("formTextError")).not.toBeNull();
            expect(screen.queryByTestId("formTextError")).toHaveTextContent('Too Short. Min length 5 symbol');
        })

        it('too long', async () => {
            renderWithRouter(<NewPassword/>)
            const password = screen.getByTestId('input_password')
            await act( async () => {
                fireEvent.change(password, {
                    target: { value: '1232133554756786788566345345345345456756878989008907890678' }
                })
            });
            await act( async () => {
                fireEvent.focusOut(password)
            });
            expect(screen.queryByTestId("formTextError")).not.toBeNull();
            expect(screen.queryByTestId("formTextError")).toHaveTextContent('Too Long. Max length 50 symbol');
        })
    })

    describe('input double password', () => {

        it('have',  () => {
            renderWithRouter(<NewPassword/>)
            expect(screen.getByTestId('input_double_password')).toBeInTheDocument()
        })

        it('validate',  async () => {
            renderWithRouter(<NewPassword/>)
            const password = screen.getByTestId('input_password')
            const double_password = screen.getByTestId('input_double_password')
            await act( async () => {
                fireEvent.change(password, {
                    target: { value: '123pass456345' }
                })
            });
            await act( async () => {
                fireEvent.change(double_password, {
                    target: { value: '123pass456345' }
                })
            });
            await act( async () => {
                fireEvent.focusOut(double_password)
            });
            expect(screen.queryByTestId("formTextError")).toBeNull();
        })

        it('different passwords',  async () => {
            renderWithRouter(<NewPassword/>)
            const password = screen.getByTestId('input_password')
            const double_password = screen.getByTestId('input_double_password')
            await act( async () => {
                fireEvent.change(password, {
                    target: { value: '123pass45634' }
                })
            });
            await act( async () => {
                fireEvent.change(double_password, {
                    target: { value: '123pass456345' }
                })
            });
            await act( async () => {
                fireEvent.focusOut(double_password)
            });
            expect(screen.queryByTestId("formTextError")).not.toBeNull();
        })

        it('too short', async () => {
            renderWithRouter(<NewPassword/>)
            const double_password = screen.getByTestId('input_double_password')
            await act( async () => {
                fireEvent.change(double_password, {
                    target: { value: 'lk' }
                })
            });
            await act( async () => {
                fireEvent.focusOut(double_password)
            });
            expect(screen.queryByTestId("formTextError")).not.toBeNull();
            expect(screen.queryByTestId("formTextError")).toHaveTextContent('Password mismatch');
        })

        it('too long', async () => {
            renderWithRouter(<NewPassword/>)
            const double_password = screen.getByTestId('input_double_password')
            await act( async () => {
                fireEvent.change(double_password, {
                    target: { value: '1232133554756786788566345345345345456756878989008907890678' }
                })
            });
            await act( async () => {
                fireEvent.focusOut(double_password)
            });
            expect(screen.queryByTestId("formTextError")).not.toBeNull();
            expect(screen.queryByTestId("formTextError")).toHaveTextContent('Password mismatch');
        })
    })

    describe('btn Send', () => {
        it('have btn Send', () => {
            renderWithRouter(<NewPassword/>)
            expect(screen.getByRole('button', { name: /Send/i} )).toBeInTheDocument()
        })
        it('not disabled', async () => {
            const { getByTestId } = renderWithRouter(<NewPassword/>)
            const btn = screen.getByRole('button', { name: /Send/i} )
            expect(btn).not.toBeDisabled()
        })
        it('disabled after error field of password', async () => {
            const { getByTestId } = renderWithRouter(<NewPassword/>)
            const password = screen.getByTestId('input_password')
            const btn = screen.getByRole('button', { name: /Send/i} )
            await act( async () => {
                fireEvent.blur(password)
            });
            expect(btn).toBeDisabled()
        })
        it('disabled after error field of double password', async () => {
            const { getByTestId } = renderWithRouter(<NewPassword/>)
            const double_password = screen.getByTestId('input_double_password')
            const btn = screen.getByRole('button', { name: /Send/i} )
            await act( async () => {
                fireEvent.blur(double_password)
            });
            expect(btn).toBeDisabled()
        })
        it('not disabled after the error disappears', async () => {
            const { getByTestId } = renderWithRouter(<NewPassword/>)
            const password = screen.getByTestId('input_password')
            const double_password = screen.getByTestId('input_double_password')
            const btn = screen.getByRole('button', { name: /Send/i} )
            await act( async () => {
                fireEvent.blur(password)
            });
            await act( async () => {
                fireEvent.change(password, {
                    target: { value: '1234f2' }
                })
            });
            await act( async () => {
                fireEvent.change(double_password, {
                    target: { value: '1234f2' }
                })
            });
            expect(btn).not.toBeDisabled()
        })
        it('clicking on a button with an empty form', async () => {
            renderWithRouter(<NewPassword/>)
            const btn = screen.getByRole('button', { name: /Send/i} )
            await act(async () => {
                btn.click()
            })
            expect(screen.getAllByTestId('formTextError').length).toBe(2)
        })
    })

    // todo-dv нужно разобраться как тестировать переход по ссылкам
    describe('link', () => {
        describe('registration', () => {
            it('to have', () => {
                renderWithRouter(<NewPassword/>)
                expect(screen.getByRole('link', {name: /registration/i}).closest('a')).toHaveAttribute('href', '/registration/')
            })
        })

        describe('forget', () => {
            it('to have',  () => {
                renderWithRouter(<NewPassword/>)
                expect(screen.getByRole('link', {name: /Authorize/i}).closest('a')).toHaveAttribute('href', '/auth/')
            })
        })
    })
})