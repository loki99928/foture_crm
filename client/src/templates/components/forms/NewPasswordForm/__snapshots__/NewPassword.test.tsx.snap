// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewPassword Snapshot 1`] = `
<DocumentFragment>
  <div
    class="blockForm"
  >
    <form
      action="#"
    >
      <div
        class="formBanner"
      >
        <img
          alt=""
          src="bg-head-form.jpg"
        />
      </div>
      <div
        class="containerFields"
      >
        <h2
          class="formTitle"
        >
          New password
        </h2>
        <div
          class="formBlockField formBlock__field"
        >
          <input
            autocomplete="on"
            class="formField"
            data-testid="input_password"
            for="password"
            id="password"
            name="password"
            type="password"
            value=""
          />
          <div
            class="formLabel"
          >
            <label
              for="password"
            >
              Your password
            </label>
          </div>
        </div>
        <div
          class="formBlockField formBlock__field"
        >
          <input
            autocomplete="on"
            class="formField"
            data-testid="input_double_password"
            for="double_password"
            id="double_password"
            name="double_password"
            type="password"
            value=""
          />
          <div
            class="formLabel"
          >
            <label
              for="double_password"
            >
              Confirm password
            </label>
          </div>
        </div>
        <div>
          <button
            class="formBtn formBtn_hover formBtn_active"
            type="submit"
          >
            Send
          </button>
        </div>
        <div
          class="formFooter footer__form"
        >
          <a
            class=""
            href="/auth/"
          >
            Authorize
          </a>
          <a
            class=""
            href="/registration/"
          >
            registration
          </a>
        </div>
      </div>
    </form>
  </div>
</DocumentFragment>
`;
